pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,222 @@
+		labelsArr = append(labelsArr, srcLabel)
+	} else {
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumClusterwideNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	}
+
+	return labelsArr
+}
+
+// read cilium network policy yaml file and convert to policy object and
+// add rules to policy engine.
+func (p *policyWatcher) addToPolicyEngine(cnpFilePath string) error {
+	// read from file and convert to cnp object
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	fileName := filepath.Base(cnpFilePath)
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		p.config.StaticCNPPath,
+		fileName,
+	)
+
+	// convert to rules
+	rules, err := cnp.Parse()
+	if err != nil {
+		return err
+	}
+
+	// update labels
+	lbls := getLabels(fileName, cnp)
+	for _, r := range rules {
+		r.Labels = lbls
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceByResource: true,
+		Source:            source.Directory,
+		Resource:          resourceID,
+	})
+
+	if err == nil {
+		p.fileNameToCnpCache[fileName] = cnp
+	}
+
+	return err
+}
+
+func (p *policyWatcher) deleteFromPolicyEngine(cnpFilePath string) error {
+	fileName := filepath.Base(cnpFilePath)
+	cnp := p.fileNameToCnpCache[fileName]
+	if cnp == nil {
+		p.log.WithField("file", fileName).Error("BUG: Policy deletion request for file which was never added")
+		return nil
+	}
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		p.config.StaticCNPPath,
+		fileName,
+	)
+	_, err := p.policyManager.PolicyDelete(getLabels(fileName, cnp), &policy.DeleteOptions{
+		Source:           source.Directory,
+		DeleteByResource: true,
+		Resource:         resourceID})
+
+	delete(p.fileNameToCnpCache, fileName)
+	return err
+}
+
+func (p *policyWatcher) watchDirectory(ctx context.Context) {
+	go func() {
+		p.log.Info("Directory policy watcher started")
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			p.log.WithError(err).Error("Initializing NewWatcher failed")
+			return
+		}
+		defer watcher.Close()
+
+		dir := p.config.StaticCNPPath
+		err = watcher.Add(dir)
+		if err != nil {
+			p.log.WithError(err).WithField(logfields.Path, dir).Error("Failed to watch policy directory. Policies will not be loaded from disk")
+			return
+		}
+
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			p.log.WithError(err).WithField("Directory", dir).Warn("Failed to read policy directory")
+			return
+		}
+
+		// Read existing cnp files before watcher was added
+		for _, f := range files {
+			absPath := filepath.Join(dir, f.Name())
+			err := p.addToPolicyEngine(absPath)
+			if err != nil {
+				p.log.WithError(err).WithField(logfields.Path, absPath).Error("Failed to add network policy to policy engine")
+			}
+			reportCNPChangeMetrics(err)
+		}
+		close(p.readStatus)
```

Comment:
A comment here, not necessarily action required:

This is only closed if we successfully read all the files. What if we exited early due to one of the error cases above (e.g. because `os.ReadDir(dir)` or `watcher.Add(dir)` failed?

As is, this will block the Daemon from ever becoming ready, which is fail-safe and makes sense given the intent of this feature (i.e. non-overridable policies). 

However, if we're blocking the daemon from becoming ready, I wonder if instead we want to use `p.log.Fatal` above to make it clear that we consider it a hard failure if policies can't be loaded from disk at all.

Reply:
make sense.. i will add Fatal for for `fsnotify.NewWatcher()`